clock/scheduledJobs/queueStatsActivityLogs.js

```diff
@@ -0,0 +1,234 @@
+ *
+ * @param {array} qt - array of queue timings
+ * @returns {object} stats - statistics based on queue timing array
+ *     i.e median, average and 95th Percentile
+ */
+function calculateStats(qt) {
+  let median = 0;
+  let average = 0;
+  let percentile95th = 0;
+  const stats = {};
+  const length = qt.length;
+
+  // sort the queue time array
+  qt.sort((a, b) => a - b);
+
+  // calculate sum of queue time
+  const sum = qt.reduce((a, b) => a + b, ZERO);
+
+  // calculate middle of array
+  const middle = Math.floor(length / TWO);
+  average = sum / length;
+
+  // calculate median
+  if (length % TWO) {
+    median = qt[middle];
+  } else {
+    median = (qt[middle - ONE] + qt[middle]) / FLOAT_TWO;
+  }
+
+  // calculate 95th Percentile
+  const index = Math.round(PERCENTILE_95TH * length);
+  percentile95th = qt[index - ONE];
+
+  // construct stats object
+  stats.averageQueueTimeMillis = average.toFixed(TWO) + 'ms';
+  stats.medianQueueTimeMillis = median.toFixed(TWO) + 'ms';
+  stats.queueTimeMillis95th = percentile95th + 'ms';
+  stats.queueTimeMillis95thNumber = percentile95th;
+
+  return stats;
+}
+
+/**
+ * Construct log object
+ *
+ * @param {object} qStats - queue time object from redis
+ * @returns {object} queueStats - queue time stats log object
+ */
+function constructLogObject(qStats) {
+  const queueStats = {};
+
+  queueStats.jobCount = qStats.jobCount;
+  queueStats.recordCount = qStats.recordCount;
+  const queueTimeArray = stringToArray(qStats.queueTimeArray);
+  queueStats.timeStamp = qStats.timeStamp;
+
+  // Calculate stats based on queue timings
+  const stats = calculateStats(queueTimeArray);
+
+  queueStats.medianQueueTimeMillis = stats.medianQueueTimeMillis;
+  queueStats.averageQueueTimeMillis = stats.averageQueueTimeMillis;
+  queueStats.queueTimeMillis95th = stats.queueTimeMillis95th;
+  queueStats.queueTimeMillis95thNumber = stats.queueTimeMillis95thNumber;
+
+  return queueStats;
+}
+
+/**
+ * Update recordCount, jobCount and push queue time when
+ * job is done.
+ *
+ * @param {integer} rc - record count for individual job
+ * @param {integer} qt - queue time for individual job
+ */
+function update(rc, qt) {
+  const timestamp = createTimeStamp();
+  const key = 'queueStats.' + timestamp;
+
+  client.hgetallAsync(key).then((reply) => {
+    if (reply) {
+      client.hmset(key, {
+        jobCount: Number(reply.jobCount) + ONE,
+        recordCount: Number(reply.recordCount) + rc,
+        queueTimeArray: addToArray(reply.queueTimeArray, qt),
+      });
+    } else {
+      client.hmset(key, {
+        jobCount: ONE,
+        recordCount: rc,
+        queueTimeArray: addToArray('', qt),
+        timeStamp: timestamp,
+      });
+    }
+  })
+  .catch((err) => {
+    throw new Error(err);
+  });
+}
+
+/**
+ * Execute the call to write queue stats activity logs.
+ */
+function execute() {
+  // Get queueStats from redis
+  client.keysAsync('queueStats*').then((keys) => {
```

Comment:
We can't call keys with wildcard in production. Please consider tracking the name of each key in a separate set.